### ISSUE TITLE
clang CUDA: shared variables sometimes are not updated (on Pascal architecture)

### ISSUE THREAD

**Issue by AuroraPerego** [llvm/llvm-project#59632]
Shared variables sometimes seem not to be updated (unless other operations are done on them forcing their reading). 
**This behaviour has been observed on a NVIDIA GeForce GTX 1080 Ti (sm_61) and on a Tesla P100 (sm_60), but not on newer GPUs. (Older GPUs have not been tested).**

To reproduce (the github repo is [here](https://github.com/AuroraPerego/clangSharedVar)):
```bash
git clone https://github.com/AuroraPerego/clangSharedVar
cd clangSharedVar
make environment
source env.sh
make
```
This will trigger an assert. Now go to `radiSort.h` and uncomment lines 181-182 and 206-207. 
After this change do:
```bash
make clean
make
./test
```
This will compile and run fine (and print a bunch of numbers).

After playing around a bit we found out that there are other ways to obtain correct results:
- replace the `__syncthreads()` at lines 180-205 with `__threadfence_block()`
- remove the `__syncthreads()` at lines 180-205 
- do an `AtomicAdd(&sharedVar,0)`

The shared variables involved are `ibs` and `p`. Since only the first thread writes on them, it seems that the other threads don't understand that the values have changed and don't read them.

clang version: clang++ 12 to 15 have been used
CUDA version: 11.5 

cc: @fwyzard

**Comment by Artem-B** [comment]
Can you try using `__barrier_sync(0)` instead of `__syncthreads`?

There was a bug #54851  in LLVM which sometimes incorrectly ignored effects of GPU threading, so this may be a duplicate of a known issue.

**Comment by Artem-B** [comment]
Also see #58626

**Comment by Artem-B** [comment]
I can not reproduce the issue on sm_60 with clang 054931dbf11687cb95576690d8f74ffd9b0da007 and CUDA-12:

```
$ ./test
Will sort 2560 signed 'ints' of size 4 using 4 significant bytes
launch for 2395
done for 2395
```

**Comment by AuroraPerego** [comment]
> Can you try using `__barrier_sync(0)` instead of `__syncthreads`?

Sorry for the delay, I tried and it works fine with `__barrier_sync(0)`

**Comment by Artem-B** [comment]
Fixed by https://github.com/llvm/llvm-project/commit/9dc7da3f9cb48ed7ba6c4806f4519d997f6671b5 https://github.com/llvm/llvm-project/issues/54851

**Comment by fwyzard** [comment]
@Artem-B unfortunately I still see the same problem with clang 16.0.5, as well as clang https://github.com/llvm/llvm-project/commit/054931dbf11687cb95576690d8f74ffd9b0da007:

```bash
# checkout LLVM and clang 16.0.5
git clone https://github.com/llvm/llvm-project.git
cd llvm-project
git checkout llvmorg-16.0.5

# configure and build clang
mkdir ../llvm-16.0.5
cd ../llvm-16.0.5
cmake -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -DCUDA_SDK_ROOT_DIR=/usr/local/cuda-11.8  -G "Unix Makefiles" ../llvm-project/llvm
make -j`nproc`

bin/clang++ --version
clang version 16.0.5 (https://github.com/llvm/llvm-project.git 185b81e034ba60081023b6e59504dfffb560f3e3)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /afs/cern.ch/user/f/fwyzard/workspace/llvm-16.0.5/bin

# checkout the reproducer using __synchthreads
cd ..
git clone https://github.com/AuroraPerego/clangSharedVar
cd clangSharedVar
git checkout 4903457e3d1a021d239da50db657377d6972a682

# build the reproducer for a P100
../llvm-16.0.5/bin/clang++ -std=c++17 -g -O2 --cuda-path=/usr/local/cuda-11.8 --offload-arch=sm_60 test.cu -L/usr/local/cuda-11.8/lib64 -lcudart -lcuda -lrt -pthreads -o test

clang-16: warning: CUDA version 11.8 is only partially supported [-Wunknown-cuda-version]

# run the reproducer on a P100
./test
```

I still get the assertion failed messages and the eventual crash:
```
Will sort 2560 signed 'ints' of size 4 using 4 significant bytes
launch for 2395
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [7,0,0], thread: [163,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [72,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [9,0,0], thread: [8,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [1,0,0], thread: [76,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [6,0,0], thread: [5,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [42,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [174,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [46,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [8,0,0], thread: [79,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [202,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [211,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [8,0,0], thread: [211,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [163,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [180,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [8,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [17,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [7,0,0], thread: [53,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [200,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [201,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [213,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [149,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [24,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [9,0,0], thread: [157,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [29,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [3,0,0], thread: [157,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [87,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [95,0,0] Assertion `c[bin] >= 0` failed.
terminate called after throwing an instance of 'std::runtime_error'
  what():
test.cu, line 125:
cudaCheck(cudaMemcpy(ind, ind_d, 2 * N, cudaMemcpyDeviceToHost));
cudaErrorAssert: device-side assert triggered

Aborted (core dumped)
```

Using clang built from https://github.com/llvm/llvm-project/commit/054931dbf11687cb95576690d8f74ffd9b0da007
```bash
../llvm-054931d/bin/clang++ -std=c++17 -g -O2 --cuda-path=/usr/local/cuda-11.8 --offload-arch=sm_60 test.cu -L/usr/local/cuda-11.8/lib64 -lcudart -lcuda -lrt -pthreads -o test

clang-17: warning: CUDA version 11.8 is only partially supported [-Wunknown-cuda-version]
```
I get the same result:
```
Will sort 2560 signed 'ints' of size 4 using 4 significant bytes
launch for 2395
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [7,0,0], thread: [163,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [72,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [9,0,0], thread: [8,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [1,0,0], thread: [76,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [6,0,0], thread: [5,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [42,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [174,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [8,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [17,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [8,0,0], thread: [79,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [46,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [8,0,0], thread: [211,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [202,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [211,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [7,0,0], thread: [53,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [200,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [201,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [213,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [163,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [2,0,0], thread: [180,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [0,0,0], thread: [149,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [24,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [4,0,0], thread: [29,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [9,0,0], thread: [157,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [3,0,0], thread: [157,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [87,0,0] Assertion `c[bin] >= 0` failed.
./radixSort.h:177: void radixSortImpl(const T *__restrict, uint16_t *, uint16_t *, uint32_t, RF) [T = int, NS = 4, RF = void (*)(const int *, unsigned short *, unsigned short *, unsigned int)]: block: [5,0,0], thread: [95,0,0] Assertion `c[bin] >= 0` failed.
terminate called after throwing an instance of 'std::runtime_error'
  what():  
test.cu, line 125:
cudaCheck(cudaMemcpy(ind, ind_d, 2 * N, cudaMemcpyDeviceToHost));
cudaErrorAssert: device-side assert triggered

Aborted (core dumped)
```

This is on a machine with Alma Linux 9.2, GCC 11.3.1, CUDA 11.8 with NVIDIA drivers 520.61.05   , and an NVIDIA P100:
```bash
nvidia-smi -L
GPU 0: Tesla P100-SXM2-16GB (UUID: GPU-ce719f31-903b-fd6c-c236-2c658455a73e)
```

I cannot easily test CUDA 12 on this machine (the current drivers are not compatible with it, and it will take some time to update the system) - but the version of CUDA should not affect the PTX generated by clang, should it ?

**Comment by fwyzard** [comment]
By the way, as @AuroraPerego already reported, replacing `__syncthreads()` with `__barrier_sync(0)` does work around the issue.

**Comment by fwyzard** [comment]
Hi @Artem-B, did you have any luck looking into this ?

We suspect that the same issue is also affecting the Intel SYCL/oneAPI compiler.

**Comment by Artem-B** [comment]
Can you check if clang built after [1ee4d880e8760256c606fe55b7af85a4f70d006d](https://github.com/llvm/llvm-project/commit/1ee4d880e8760256c606fe55b7af85a4f70d006d) makes any difference?

**Comment by fwyzard** [comment]
Using a recent build, the reproducer now passes the tests.

I will try to check the full application.

**Comment by Artem-B** [comment]
If you could check PTX generated by clang for the reproducer that would help to confirm whether the root cause was indeed the thread mis-convergence worked around by [1ee4d880e8760](https://github.com/llvm/llvm-project/commit/1ee4d880e8760256c606fe55b7af85a4f70d006d)

If you do see an `exit`  instruction in the PTX generated by the recent clang, that would be the place where the old PTX would previously confuse ptxas.  LLVM would consider that code unreachable (and thus control flow branch terminated), but for ptxas that would look like a continuous code execution falling through to the next instruction that would belong to a completely different control flow branch. With an explicit `exit`, ptxas can now re-converge the threads properly.

A word of caution -- the code LLVM ends up considering `unreachable` may be a result of undefined behavior in the user code, so the original misconvergence issue may have been a secondary failure. I would recommend mapping back the `exit` instruction in the PTX back to the source code (compiling with -g1 will generate .line directives to help with that) and double-checking that it is indeed expected to be unreachable.

**Comment by fwyzard** [comment]
hi @Artem-B, I've bisected the `main` branch between January 1st (when the reproducer was failing) and June 1st (when it was working), and I think I found the commit that fixed the problem around February 28th:

...
 :heavy_check_mark: 4668dd8c4ab1 [libc++] Run clang-tidy in all configurations that are run in the Docker container
 :heavy_check_mark: 2c1433453d16 [flang] Block construct
 :heavy_check_mark: fbec8f04ab39 [SimplifyCFG] Do not hoist/sink convergent function calls
 :x: 83c0d6524805 [gn build] Manually port D140588
 :x: 853059a15011 [libc++] Addresses LWG3782.
...

From the description of fbec8f04ab39

> **[SimplifyCFG] Do not hoist/sink convergent function calls**
> 
> Currently SimplifyCFG hoists/sink common instructions in then/else basic blocks
when certain options are enabled, which is the case for default clang optimization
pipelines for -O3. It tries to hoist/sink convergent function calls in divergent
control flow, which causes incorrect ISA generated for GPU, e.g.
https://github.com/ROCm-Developer-Tools/HIP/issues/3172
> 
> This patch fixes that by conservatively disable hoisting/sinking common
convergent function calls in then/else blocks.

it makes sense that it fixed the issue we were seeing.
